File: week12.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

// Course class
class Course {
public:
    int code;
    string name;
    vector<class Student*> students;  // who is taking this course

    Course(int c, string n) {
        code = c;
        name = n;
    }

    void addStudent(Student* s) {
        students.push_back(s);
    }
};

// Student class
class Student {
public:
    int id;
    string name;
    vector<Course*> courses;

    Student(int i, string n) {
        id = i;
        name = n;
    }

    void takeCourse(Course* c) {
        courses.push_back(c);
        c->addStudent(this);
    }
};

// Registration system
class RegSystem {
public:
    void registerStudent(Student* s, Course* c) {
        cout << s->name << " is registering for " << c->name << endl;
        s->takeCourse(c);
    }
};

int main() {
    // Students
    Student s1(1, "Maria");
    Student s2(2, "Ezra");

    // Courses
    Course c1(2210, "Programming 2");
    Course c2(2270, "Calculus 3");

    // Registration
    RegSystem reg;
    reg.registerStudent(&s1, &c1);
    reg.registerStudent(&s2, &c1);
    reg.registerStudent(&s1, &c2);

    // Show enrolled courses
    Student* allStudents[] = { &s1, &s2 };
    for (int i = 0; i < 2; i++) {
        cout << allStudents[i]->name << " is taking ";
        for (int j = 0; j < allStudents[i]->courses.size(); j++) {
            cout << allStudents[i]->courses[j]->name;
            if (j != allStudents[i]->courses.size() - 1) {
                cout << ", ";
            }
        }
        cout << endl;
    }

    return 0;
}

File: week12.py
# Student class
class Student:
    def __init__(self, id, name):
        self.id = id
        self.name = name
        self.courses = []

    def take_course(self, course):
        self.courses.append(course)
        course.add_student(self)

# Course class
class Course:
    def __init__(self, code, name):
        self.code = code
        self.name = name
        self.students = []

    def add_student(self, student):
        if student not in self.students:
            self.students.append(student)

# Registration system
class RegistrationSystem:
    def register(self, student, course):
        print(student.name + " is registering for " + course.name)
        student.take_course(course)

# Example
s1 = Student(1, "Maria")
s2 = Student(2, "Ezra")

c1 = Course(2210, "Programming 2")
c2 = Course(2270, "Calculus 3")

reg = RegistrationSystem()
reg.register(s1, c1)
reg.register(s2, c1)
reg.register(s1, c2)

# Show which courses each student has
for s in [s1, s2]:
    course_names = []
    for c in s.courses:
        course_names.append(c.name)
    print(s.name + " is taking " + str(course_names))
